OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs

```diff
@@ -39,16 +39,40 @@ public class AttackGarrisonedInfo : AttackFollowInfo, Requires<CargoInfo>
 		[Desc("Fire port yaw cone angle.")]
 		public readonly WAngle[] PortCones = null;
 
+		public FirePort[] Ports { get; private set; }
+
 		[PaletteReference] public readonly string MuzzlePalette = "effect";
 
 		public override object Create(ActorInitializer init) { return new AttackGarrisoned(init.Self, this); }
+		public void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			if (PortOffsets.Length == 0)
+				throw new YamlException("PortOffsets must have at least one entry.");
+
+			if (PortYaws.Length != PortOffsets.Length)
+				throw new YamlException("PortYaws must define an angle for each port.");
+
+			if (PortCones.Length != PortOffsets.Length)
+				throw new YamlException("PortCones must define an angle for each port.");
+
+			var p = new List<FirePort>();
```

Comment:
Minor thing since you're moving this around: could you do ```Ports = new FirePort[PortOffsets.Length]``` and write into the array directly.